connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -338,17 +342,24 @@ private void processTargetStateChanges(Set<String> connectorTargetStateChanges)
     public void halt() {
         synchronized (this) {
             // Clean up any connectors and tasks that are still running.
-            log.info("Stopping connectors and tasks that are still assigned to the worker");
-            worker.stopConnectors();
-            worker.stopAndAwaitTasks();
+            log.info("Stopping connectors and tasks that are still assigned to this worker.");
+            List<Callable<Void>> callables = new ArrayList<>();
+            for (String connectorName : new ArrayList<>(worker.connectorNames())) {
+                callables.add(getConnectorCallable(connectorName, ConnectorStatus.State.DESTROYED));
+            }
+            for (ConnectorTaskId taskId : new ArrayList<>(worker.taskIds())) {
+                callables.add(getTaskCallable(taskId, TaskStatus.State.DESTROYED));
+            }
+            startAndStop(callables);
 
             member.stop();
 
             // Explicitly fail any outstanding requests so they actually get a response and get an understandable reason
             // for their failure
-            while (!requests.isEmpty()) {
-                HerderRequest request = requests.poll();
+            HerderRequest request = requests.pollFirst();
```

Comment:
There are several reasons why using an enhanced `for` loop with a concurrent collection here is not what we want. 

Case 1: Between the end of the foreach loop and the call to `clear` a request is added (a call to the REST API served by another thread adds a request). This request's callback will never be called. It will be cleared although it wasn't traversed through the iterator used by the foreach loop. 

Case 2: Even worse. Requests added while the foreach loop executes might not be seen either. As mentioned in the javadocs [here](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentSkipListMap.html#keySet()), an iterator to a concurrent collection as the one used here, is a weakly consistent iterator, and although it won't throw an exception due to concurrent modification of the collection, it won't necessarily reflect the modifications either: It _"guarantees to traverse elements as they existed upon construction of the iterator, and may (but is not guaranteed to) reflect any modifications subsequent to construction"_